storage/src/main/java/org/apache/kafka/server/purgatory/DelayedRemoteListOffsets.java

```diff
@@ -0,0 +1,200 @@
+                                    Map<TopicPartition, ListOffsetsPartitionStatus> statusByPartition,
+                                    Consumer<TopicPartition> partitionOrException,
+                                    Consumer<List<ListOffsetsResponseData.ListOffsetsTopicResponse>> responseCallback) {
+        super(delayMs);
+        this.version = version;
+        this.statusByPartition = statusByPartition;
+        this.partitionOrException = partitionOrException;
+        this.responseCallback = responseCallback;
+        // Mark the status as completed, if there is no async task to track.
+        // If there is a task to track, then build the response as REQUEST_TIMED_OUT by default.
+        statusByPartition.forEach((topicPartition, status) -> {
+            status.completed(status.futureHolderOpt().isEmpty());
+            if (status.futureHolderOpt().isPresent()) {
+                status.responseOpt(Optional.of(buildErrorResponse(Errors.REQUEST_TIMED_OUT, topicPartition.partition())));
+            }
+            LOG.trace("Initial partition status for {} is {}", topicPartition, status);
+        });
+    }
+
+    /**
+     * Call-back to execute when a delayed operation gets expired and hence forced to complete.
+     */
+    @Override
+    public void onExpiration() {
+        statusByPartition.forEach((topicPartition, status) -> {
+            if (!status.completed()) {
+                LOG.debug("Expiring list offset request for partition {} with status {}", topicPartition, status);
+                status.futureHolderOpt().ifPresent(futureHolder -> futureHolder.jobFuture().cancel(true));
+                recordExpiration(topicPartition);
+            }
+        });
+    }
+
+    /**
+     * Process for completing an operation; This function needs to be defined
+     * in subclasses and will be called exactly once in forceComplete()
+     */
+    @Override
+    public void onComplete() {
+        Map<String, List<ListOffsetsResponseData.ListOffsetsPartitionResponse>> groupedByTopic = new HashMap<>();
+        for (Map.Entry<TopicPartition, ListOffsetsPartitionStatus> entry : statusByPartition.entrySet()) {
+            List<ListOffsetsResponseData.ListOffsetsPartitionResponse> partitions =
+                    groupedByTopic.computeIfAbsent(entry.getKey().topic(), k -> new ArrayList<>());
+            if (entry.getValue().responseOpt().isPresent()) {
+                partitions.add(entry.getValue().responseOpt().get());
+            }
+        }
+        List<ListOffsetsResponseData.ListOffsetsTopicResponse> response = new ArrayList<>();
+        for (Map.Entry<String, List<ListOffsetsResponseData.ListOffsetsPartitionResponse>> entry : groupedByTopic.entrySet()) {
+            response.add(new ListOffsetsResponseData.ListOffsetsTopicResponse()
+                    .setName(entry.getKey())
+                    .setPartitions(entry.getValue()));
+        }
+        responseCallback.accept(response);
```

Comment:
Could you please consider changing the type from `List<ListOffsetsResponseData.ListOffsetsTopicResponse>` to `Collection<ListOffsetsResponseData.ListOffsetsTopicResponse>`, allowing us to streamline the code as follow:

```java
        Map<String, ListOffsetsResponseData.ListOffsetsTopicResponse> another = new HashMap<>();
        statusByPartition.forEach((tp, status) -> {
            var r = another.computeIfAbsent(tp.topic(), n -> new ListOffsetsResponseData.ListOffsetsTopicResponse());
            status.responseOpt().ifPresent(res -> r.partitions().add(res));
        });
        responseCallback.accept(another.values());
```

Reply:
Good idea, this can save us iterating over the responses twice.